lib/ClimaCoreSpectra/src/gcm_spectra.jl

```diff
@@ -0,0 +1,392 @@
+# Main implentation based on: https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Common/Spectra/power_spectrum_gcm.jl
+
+#=
+Cleanup items:
+
+ - Inherit array types (i.e., don't assume things are `Array`, as we may have `CuArray` for gpu down the road), instead use `Arr where {Arr <: AbstractArray}`
+ - Follow type promotion rules: https://docs.julialang.org/en/v1/manual/conversion-and-promotion/
+ - Do we need this to be allocation-free?
+ - Can we use external packages (e.g., RootSolvers) for loops that seem to be performing iterations?
+ - Improve documentation
+
+ - Add main module string with todo list:
+  - What API functions are needed?
+    - What are the function argument types?
+    - What do they do / output, in equation form?
+=#
+
+"""
+    AbstractSpectralSphericalMesh
+
+An abstract spherical mesh data structure for calculating spectra.
+"""
+abstract type AbstractSpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4} end
+
+"""
+    SpectralSphericalMesh
+
+Spherical mesh data structure for calculating spectra. The mesh represents a regular lat-long grid.
+"""
+mutable struct SpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4} <:
+               AbstractSpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4}
```

Comment:
```suggestion
struct SpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4} <:
               AbstractSpectralSphericalMesh{FT, ArrF3, ArrI2, ArrC3, ArrC4}
```

Reply:
On a second thought, the way this is set up now, it actually needs to be mutable. We mute one of its arguments later on. 
```
mesh.qnm = compute_legendre!(
        FT,
        mesh.num_fourier,
        mesh.num_spherical,
        sinθ,
        mesh.nθ,
    )
```
I would leave this to following clean-up PRs, as that `compute_legendre` would most likely be substitute by the AssociatedLegendrePolynomials.jl package functions